### PR TITLE
newton element

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
newton's third law, shoot a gun and go flying

## Why It's Good For The Game
memes

## Changelog
:cl:
add: buckling yourself to a roller chair and shooting now moves you
/:cl:

### CODE REVIEW

**Comment by Kuro-ai** [comment]
SPEED MERGE PLS

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -747,42 +747,9 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
 
 	playsound(user.loc, 'sound/effects/extinguish.ogg', 52, 1, 7)
 
-	var/direction = get_dir(user,target)
+	SEND_SIGNAL(user, COMSIG_MOB_EXTINGUISHER_USE, target, 4)
```
Comment (Rohesie) [inline code/game/objects/items.dm]: Put this before the sound and feedback, a couple of lines up.

**Comment by Hughgent** [comment]
You mentioned in the Discord that it's 1 bullet per tile. If you could make that 3 bullets per tile, for most rifles bursts, I think that would be 'slightly' more realistic.

File: code/datums/elements/newton.dm [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+/datum/element/newton/proc/propel(mob/M, direction, force)
+	if(!M.buckled || !isobj(M.buckled) || M.buckled.anchored)
+		Detach(M)
+
+	var/rep = CLAMP((4-force)*2, 0, 8)
+
+	move_chair(M.buckled, reverse_direction(direction), rep)
+
+/datum/element/newton/proc/move_chair(obj/B, movementdirection, repetition=0)
+	step(B, movementdirection)
```
Comment (Rohesie) [inline code/datums/elements/newton.dm]: Should use something like `relaymove()`

File: code/game/objects/items.dm (hunk 1 above)
Comment (spookydonut) [reply code/game/objects/items.dm]: Same place as the movement code was before.

**Comment by spookydonut** [comment]
> 
> 
> You mentioned in the Discord that it's 1 bullet per tile. If you could make that 3 bullets per tile, for most rifles bursts, I think that would be 'slightly' more realistic.

not without making it a lot more complicated

File: code/__DEFINES/components.dm [hunk 3]
```diff
@@ -161,7 +161,7 @@
 	#define COMPONENT_ITEM_CLICKCTRLON_INTERCEPTED (1<<0)				//from base of mob/CtrlClickOn(): (/atom, /mob)
 
 // /obj/item/weapon/gun signals
-#define COMSIG_GUN_FIRE "gun_fire"
+#define COMSIG_GUN_FIRE "gun_fire"								//:() (target, user)
```
Comment (ninjanomnom) [inline code/__DEFINES/components.dm]: Do the actual comment format

File: code/game/objects/items.dm (hunk 1 above)
Comment (ninjanomnom) [reply code/game/objects/items.dm]: This isnt movement code though, the signal should be as early in the proc as reasonable by default

File: code/datums/elements/newton.dm [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+/datum/element/newton/proc/fire_gun(datum/source, atom/target, obj/item/weapon/gun/gun, mob/living/user)
+	propel(user, get_dir(user, target), 0)
+
+/datum/element/newton/proc/propel(mob/M, direction, force)
+	if(!M.buckled || !isobj(M.buckled) || M.buckled.anchored)
+		Detach(M)
+
+	var/rep = CLAMP((4-force)*2, 0, 8)
+
+	move_chair(M.buckled, reverse_direction(direction), rep)
```
Comment (ninjanomnom) [inline code/datums/elements/newton.dm]: This seems very strange to be adding to the mob whenever they sit on a chair, not only that you're assuming here that the chair exists. You should instead either make the element go on the chair and use a signal for when mobs buckle or make the element more generic.

If it *is* generic you probably want to make it support going on either side, and at that point you've recreated space drift.

File: code/datums/elements/newton.dm [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+/datum/element/newton/proc/move_chair(obj/B, movementdirection, repetition=0)
+	step(B, movementdirection)
```
Comment (ninjanomnom) [inline code/datums/elements/newton.dm]: You want the walk() proc

File: code/datums/elements/newton.dm [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+/datum/element/newton/proc/exinguish(datum/source, atom/target_atom, force)
+	var/mob/M = source
+
+	propel(M, get_dir(M, target_atom), force)
+
+/datum/element/newton/proc/throw_item(datum/source, atom/target)
+	var/mob/M = source
+
+	propel(M, get_dir(M, target), 0)
```
Comment (ninjanomnom) [inline code/datums/elements/newton.dm]: You dont need to typecast

File: code/datums/elements/newton.dm [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	UnregisterSignal(target, list(
+		COMSIG_MOB_EXTINGUISHER_USE, 
+		COMSIG_HUMAN_ITEM_THROW, 
+		COMSIG_HUMAN_GUN_FIRED, 
+		COMSIG_HUMAN_GUN_AUTOFIRED))
+
+/datum/element/newton/proc/exinguish(datum/source, atom/target_atom, force)
+	var/mob/M = source
+
+	propel(M, get_dir(M, target_atom), force)
```
Comment (ninjanomnom) [inline code/datums/elements/newton.dm]: ditto

File: code/game/objects/objs.dm [hunk 8]
```diff
@@ -201,13 +203,15 @@
...
 /obj/proc/do_buckle(mob/M, mob/user, silent = FALSE)
+	SHOULD_CALL_PARENT(1)
 	if(!silent)
 		send_buckling_message(M, user)
 	M.buckled = src
 	M.loc = src.loc
 	M.setDir(dir)
 	M.update_canmove()
 	src.buckled_mob = M
+	M.AddElement(/datum/element/newton)
```
Comment (ninjanomnom) [inline code/game/objects/objs.dm]: Not all bucklable objects should allow drift

File: code/datums/elements/newton.dm (hunk 5 above)
Comment (spookydonut) [reply code/datums/elements/newton.dm]: I ripped this from tg which uses step

Comment (ninjanomnom) [reply code/datums/elements/newton.dm]: then they do it wrong

Comment (ninjanomnom) [reply code/datums/elements/newton.dm]: then they do it wrong

**Pull request closed without merging**